[WWST-6515] - New DTH for the Fibaro Walli Double Switch, device preferences included

KKlimczukS opened the pull request:
@tpmanley @greens @dkirker @PKacprowiczS @ZWozniakS @MGoralczykS @MWierzbinskaS 

[WWST-6515] - New DTH for the Fibaro Walli Double Switch
**-> based on "Z-Wave Multi Metering Switch"**
**-> includes support for device preferences**

There are some device preferences that are commented intentionally (to be deleted later) - I just wanted to ask You guys to double check it if we should get rid of those ones.

**-> command "reset" was left intentionally**
I suggested (ICP-12599) adding a dedicated button in OneApp to reset power consumption values on all endpoints.

Please, take a look at the code.

KKlimczukS committed:
New DTH for the Fibaro Walli Double Switch, device preferences included

greens commented:
I appreciate what you're doing here, but I think it's too much, especially for a switch. I think it would be better to just put this into `multi-metering-switch`.

Also if you generated these preferences using @PKacprowiczS 's tool it looks like there is a bug in it around spacing for the preference description.

PKacprowiczS commented:
@greens I think issue about spacing is caused somewhere in app and the fact that text isn't wrapped in a bounding box, not by description definition itself.

KKlimczukS commented:
> I appreciate what you're doing here, but I think it's too much, especially for a switch. I think it would be better to just put this into `multi-metering-switch`.
@greens Please, take a look at JIRA-> WWST-6515

KKlimczukS commented:
@greens @dkirker Can we continue with this one ?

greens reviewed (changes requested):
you have a typo in the filename, and it would make more sense to have this under the Fibaro namespace/folder.

greens commented on devicetypes/smartthings/fibaro-double-switch.src/fibaro-doble-switch.groovy:
```diff
@@ -0,0 +1,799 @@
+	def model = Integer.toHexString(cmd.productId)
+	updateDataValue("mfr", mfr)
+	updateDataValue("model", model)
+}
```
none of this is necessary anymore

KKlimczukS replied:
ok, fixed

greens commented on devicetypes/smartthings/fibaro-double-switch.src/fibaro-doble-switch.groovy:
```diff
@@ -0,0 +1,799 @@
+	log.debug "Configure..."
+	response([
+			encap(zwave.multiChannelV3.multiChannelEndPointGet()),
+			encap(zwave.manufacturerSpecificV2.manufacturerSpecificGet())
```
not necessary

KKlimczukS replied:
ok, fixed

greens commented on devicetypes/smartthings/fibaro-double-switch.src/fibaro-doble-switch.groovy:
```diff
@@ -0,0 +1,799 @@
+		description: "This parameter determines how the device will react in the event of power supply failure (e.g. power outage). The parameter is not relevant for outputs set to pulse mode (parameter 150/151 set to 2)."
+	],
+	[
+		name: "LED frame � power limit", key: "ledFrame�PowerLimit", type: "range",
```
fix the spaces in these names

KKlimczukS replied:
ok, fixed

KKlimczukS commented:
> you have a typo in the filename, and it would make more sense to have this under the Fibaro namespace/folder.

@greens Did You mean to move it to "fibargroup" folder ?

KKlimczukS committed:
WWST-6515 - fixes

dkirker commented:
@KKlimczukS The filename is still incorrect. It should be:
`smartthings/fibaro-walli-double-switch.src/fibaro-walli-double-switch.groovy`

And yes, the `fibargroup` namespace.

dkirker commented on devicetypes/smartthings/fibaro-double-switch.src/fibaro-double-switch.groovy:
```diff
@@ -0,0 +1,786 @@
+	for (def endpoint : 2..numberOfSwitches) {
+		try {
+			String childDni = "${device.deviceNetworkId}:$endpoint"
+			def componentLabel = device.displayName + "-${endpoint}"
```
I think the component name would look a little cleaner as `device.displayName + " ${endpoint}"`.

dkirker replied:
There's supposed to be a space before the `$endpoint` part. It appears that the formatting didn't treat that space as non-breaking.

```
device.displayName + " ${endpoint}"
```

KKlimczukS replied:
ok, fixed

dkirker commented on devicetypes/smartthings/fibaro-double-switch.src/fibaro-double-switch.groovy:
```diff
@@ -0,0 +1,786 @@
+		 description: "This parameter determines maximum active power. Exceeding it results in the LED frame flashing violet. Function is active only when parameter 11 is set to 8 or 9."
+	],
+	[
+			name: "LED frame - colour when ON", key: "ledFrame-ColourWhenOn", type: "enum",
```
The double (and sometimes more) indenting for lists throughout the file gets a bit distracting. It would be much cleaner/more readible as just a single indent. Though this is the only instance of it in `getParameterMap()`.

KKlimczukS replied:
ok, fixed (in a whole file). Please let me know if it looks better now.

dkirker reviewed (commented):
Left a couple of comments. Otherwise, looks good. Let's move it to the `fibargroup` namespace so that it can be with the others. (Ideally, we'd get them all moved over there. But that discussion can happen later.)

KKlimczukS commented:
> smartthings/fibaro-walli-double-switch.src/fibaro-walli-double-switch.groovy


@dkirker @greens  Please, double check, I thought it should be changed to -> 
**fibargroup**/fibaro-walli-double-switch.src/fibaro-walli-double-switch.groovy

dkirker commented:
@KKlimczukS Yes, my apologies. I was originally commenting that the file name and path was missing the `walli` part and then added that the namespace should be updated to fibargroup but didn't revise the file name.

It should be `fibargroup/fibaro-walli-double-switch.src/fibaro-walli-double-switch.groovy`.

KKlimczukS committed:
WWST-6515 - fixes

KKlimczukS commented:
> @KKlimczukS Yes, my apologies. I was originally commenting that the file name and path was missing the `walli` part and then added that the namespace should be updated to fibargroup but didn't revise the file name.
> 
> It should be `fibargroup/fibaro-walli-double-switch.src/fibaro-walli-double-switch.groovy`.

@dkirker ok, fixed (renamed and moved to another folder)

KKlimczukS commented:
> you have a typo in the filename, and it would make more sense to have this under the Fibaro namespace/folder.

@greens ok, done.

KKlimczukS commented:
@dkirker @greens Please, could You also take a look at some preferences (that are commented) that are defined in this method -> getParameterMap()
They will be removed, but i need Your double check here.

KKlimczukS committed:
WWST-6515 - Constrained list of supported device settings.

KKlimczukS committed:
WWST-6515 - Corrected deviceJoinName

KKlimczukS committed:
WWST-6515, ICP-12783 - added a DTH for a child device that is in the same namespace - fibargroup, removed unnecessary comment from main DTH

greens commented on devicetypes/fibargroup/fibaro-walli-double-switch.src/child-metering-switch.groovy:
```diff
@@ -0,0 +1,74 @@
+ *
+ */
+metadata {
+	definition(name: "Child Metering Switch", namespace: "fibargroup", author: "SmartThings", mnmn: "SmartThings", vid: "generic-switch-power-energy") {
```
As far as I know you could still have just used the smartthings-namespace version of this DTH?

KKlimczukS replied:
@greens @dkirker 
You recommended moving this (parent) DTH to the "fibargroup" namespace.
After that it was noticed that child device wasn't created (reported in ICP-12783)

that's how  adding a child device looks in a parent DTH:
`addChildDevice("Child Metering Switch", childDni, device.getHub().getId(), [
				completedSetup	: true,
				label			: componentLabel,
				isComponent		: false
			])`

Is there another way (syntax) to reach out for the Child Metering Switch (that is in smartthings namespace already) ? If so - then copying the child-metering-switch will not be needed on the "fibargroup" side.

greens replied:
Ah. Yeah just looked at the underlying code. Could you use `fibaro-double-switch-2-usb`?

KKlimczukS replied:
@greens We might try that, but fibaro-double-switch-2-usb is defined in a bit different namespace "**FibarGroup**". Will it work without changing it ?
Suggestion: Maybe all those fibaro dths should be unified to work under the same namespace ? That would avoid confusions.

greens replied:
My hunch is that the namespace is case-insensitive, but test it out. The main thing we want to avoid is having to create two DTHs here, so if you can't get that to work, go ahead and move this back to the smartthings namespace.

KKlimczukS replied:
@greens 
If we want to stick to "fibaro-double-switch-2-usb" (who doesn't send its DNI when on/off is called) then:
1)  both parent "Fibaro Double Switch 2 ZW5" and child "fibaro-double-switch-2-usb"
2)  or  parent "Fibaro Walli Double Switch" (who handles child devices in more generic way right now)

 should be refactored.

KKlimczukS replied:
UPDATE: both parent DTHs were refactored, a child DTH was refactored too.

greens commented on devicetypes/fibargroup/fibaro-walli-double-switch.src/fibaro-walli-double-switch.groovy:
```diff
@@ -0,0 +1,533 @@
+			8: "colour changes smoothly depending on measured power",
+			9: "colour changes in steps depending on measured power"
+		],
+		description: "This parameter defines the LED colour when thedevice is ON. When set to 8 or 9, LED frame colour will change depending on he measured power and parameter 10. Other colours are set permanently and do not depend on power consumption."
```
`the device`, `the measured`,

KKlimczukS replied:
ok, fixed

greens commented on devicetypes/fibargroup/fibaro-walli-double-switch.src/fibaro-walli-double-switch.groovy:
```diff
@@ -0,0 +1,533 @@
+
+		command "reset"
+
+		fingerprint mfr: "010F", prod: "1B01", model: "1000", deviceJoinName: "Switch" //Fibaro Walli Double Switch FGWDSEU-221, Raw Description zw:Ls type:1001 mfr:010F prod:1B01 model:1000 ver:5.01 zwv:6.02 lib:03 cc:5E,55,98,9F,56,6C,22 sec:25,85,8E,59,86,72,5A,73,32,70,71,75,5B,7A,60 role:05 ff:9D00 ui:9D00 epc:2
```
can you move the comment to another line? this DTH is very W I D E

KKlimczukS replied:
ok, fixed

greens commented on devicetypes/fibargroup/fibaro-walli-double-switch.src/fibaro-walli-double-switch.groovy:
```diff
@@ -0,0 +1,533 @@
+		name: "LED frame - brightness", key: "ledFrame-Brightness", type: "range",
+		parameterNumber: 13, size: 1, defaultValue: 100,
+		range: "1..102",
+		description: "This parameter allows to adjust the LED frame brightness."
```
I'd change to:

"adjust the LED frame brightness"

In fact for most of these I'd drop the "this parameter defines/allows" wording

KKlimczukS replied:
ok, it will be fixed (all device settings details were taken - as they were -  from wave alliance official product data)

KKlimczukS replied:
ok, fixed

KKlimczukS committed:
WWST-6527 - fixes (deviceJoinName, device settings descriptions)

cbaumler commented on devicetypes/fibargroup/fibaro-walli-double-switch.src/fibaro-walli-double-switch.groovy:
```diff
@@ -0,0 +1,535 @@
+		if (state.currentPreferencesState."$key".status == "disablePending" && preferenceValue == preference.disableValue) {
+			state.currentPreferencesState."$key".status = "disabled"
+		} else {
+			runIn(5, "syncConfiguration", [overwrite: true])
```
Is `[overwrite: true]` necessary? @dkirker mentioned in another PR that was the default behavior.

KKlimczukS replied:
Even if that's the default behavior, I would leave [overwrite: true] - for clarity.

greens commented on devicetypes/fibargroup/fibaro-walli-double-switch.src/fibaro-walli-double-switch.groovy:
```diff
@@ -0,0 +1,535 @@
+		parameterNumber: 24, size: 1, defaultValue: 0,
+		optionInactive: 0, inactiveDescription: "default (1st button controls 1st channel, 2nd button controls 2nd channel)",
+		optionActive: 1, activeDescription: "reversed (1st button controls 2nd channel, 2nd button controls 1st channel)",
+		description: "Reversing the operation of the buttons."
```
`Reverse` here and below

KKlimczukS replied:
ok, fixed

KKlimczukS committed:
WWST-6515 - fix (device settings descriptions)

KKlimczukS committed:
WWST-6515 - refactoring of both parent DTHs so they could use the same child device DTH, refactoring of the child device DTH (sending deviceNetworkId to a parent)

greens commented:
Since both of the main DTHs here are very explicitly for double switches and have configuration options that basically make them invalid to be used for any other model of device, I think it's okay to avoid genericizing the code if it means we don't have to touch extra files.

KKlimczukS commented:
> Since both of the main DTHs here are very explicitly for double switches and have configuration options that basically make them invalid to be used for any other model of device, I think it's okay to avoid genericizing the code if it means we don't have to touch extra files.

@greens @dkirker 
I wanted to avoid that refactoring too, but it finally turned out that very minor changes had to be done in those files. Are they acceptable ?

greens commented:
@KKlimczukS You didn't need to change either existing DTH. They both currently operate on the assumption that they know what the child DNI is. You can alter your new DTH to do the same instead of writing it in a way that makes it possible for it to have more than one child switch, which is what is requiring the changes in the existing DTHs.

KKlimczukS commented:
> @KKlimczukS You didn't need to change either existing DTH. They both currently operate on the assumption that they know what the child DNI is. You can alter your new DTH to do the same instead of writing it in a way that makes it possible for it to have more than one child switch, which is what is requiring the changes in the existing DTHs.

@greens My intention was to refactor that child DTH so it could be more "opened" to work among many child devices (in fibargroup namespace) in future integrations of Fibaro devices that may have more than 2 endpoints.
I will revert those changes and refactor Fibaro Walli Double Switch DTH code only.

KKlimczukS committed:
WWST-6515 - reverting previous changes, refactoring of parent DTH

greens commented:
@KKlimczukS I understand, but the addition of preferences means that it's unlikely that any future devices will be able to be added to this DTH.

If the older Fibaro Double Switch doesn't have parameters that match up with this one, I don't expect that any future switch (with more endpoints) will either.
